service/topology_coordinator.cc

```diff
@@ -2260,9 +2261,12 @@ future<> topology_coordinator::build_coordinator_state(group0_guard guard) {
     rtlogger.info("waiting for all nodes to finish upgrade to raft schema");
     release_guard(std::move(guard));
     co_await _group0.wait_for_all_nodes_to_finish_upgrade(_as);
-    guard = co_await start_operation();
+
+    rtlogger.info("migrating system_auth keyspace data");
+    co_await auth::migrate_to_auth_v2(_sys_ks.query_processor(), _group0.client(), _as);
```

Comment:
Shouldn't this be done in one transaction? If the next one fails, we are in a half-migrated state and the next build_coordiantor_state will overwrite the tables already migrated.

Reply:
that still should be fine, failures in the migration procedure are handled already - it retries from scratch. And as I understand restarting migrate_to_auth_v2 will anyway overwrite any half-written data in the new tables.

Reply:
`migrate_to_auth_v2` internally uses `_group0.client().start_operation()`. **This is unsafe** -- it doesn't protect from coordinator failover, which `topology_coordinator::start_operation()` does (it checks the term)!

Reply:
That applies to the guard(s) taken by `announce_mutations_with_batching` too

So perhaps instead of passing client, we should be passing a function that takes the guard. Or something.

Reply:
> it retries from scratch

Ah but shouldn't we clear the data we already migrated?

I guess no because we assume that the admin does not do any changes while migration is running

Reply:
> I guess no because we assume that the admin does not do any changes while migration is running

Yes that's the assumption here.

Reply:
We plan to make the procedure fully automatic (no admin coordinator) and safe from concurrent changes so please leave a FIXME at least (and I asked @piodul to create follow-up issue where you can leave a comment too)

Reply:
I think group0_client has to be made safe because it's very weird to use client for one thing:
`group0_client.prepare_command`
and then `topology_coordinator.start_operation` for another thing while group0_client has also `start_operation` method.

Basically it's a trap people will fall 100% into.

Reply:
> We plan to make the procedure fully automatic (no admin coordination) and safe from concurrent changes so please leave a FIXME at least

basically I can call qp.execute_internal("TRUNCATE TABLE ...") that should fix it right?

Reply:
> Also btw shouldn't term_changed_error be retried on query_processor level the same as concurrent modification is?

But term_changed_error is not thrown by group0 client API. In topology coordinator code it's thrown by topology coordinator code which adds the term check on top of group 0 client api

Reply:
> I think group0_client has to be made safe because it's very weird to use client for one thing:
group0_client.prepare_command and then topology_coordinator.start_operation for another thing while group0_client has also start_operation method.

> Basically it's a trap people will fall 100% into.

I don't understand. group0_client is safe. Its provides the API to safely commit/apply a single group 0 command under a single guard.

But topology_coordinator wants additional guarantees and checks so it adds them, it's a layer built on top of group 0. Topology coordinator is dropping and re-taking guard and of course while lock is dropped, coordinator may change. So we need to check that somehow, that's why we introduced the term check

Reply:
> basically I can call qp.execute_internal("TRUNCATE TABLE ...") that should fix it right?

Not so fast, probably not, unless we implement truncate as a group 0 command. But maybe. I don't know.

Reply:
> But term_changed_error is not thrown by group0 client API. In topology coordinator code it's thrown by topology coordinator code which adds the term check on top of group 0 client api

In short:
- migration is done by topology coordinator. So it needs to preserve the checks that commands are sent by the coordinator and not some other nodes
- then regular query_processor auth statements can be safely done by any node. So they don't need these checks

Reply:
I think it's very problematic to implement, as migrate_to_auth_v2 in edge cases can obtain additional guards. Although as I understand this check will be executed anyway due to `guard = co_await start_operation();` line right after auth-v2 migration. So it should be safe, no?

Reply:
> so it should be used where exactly? In migration code only? Or everywhere i.e. auth DML , migration_manager?

Do we need to guarantee for auth DML and migration_manager statements to run on topology coordinator only, or are they allowed to be executed on any node?

Reply:
> I think it's very problematic to implement, as migrate_to_auth_v2 in edge cases can obtain additional guards. 

It doesn't seem that problematic at a glance, you can provide a function that obtains a guard as a parameter to the functions which need to take a guard

> Although as I understand this check will be executed anyway due to guard = co_await start_operation(); line right after auth-v2 migration. So it should be safe, no?

I don't think so. Suppose coordinator failover happens, and the first coordinator is unaware and keeps performing the migration. The second coordinator finishes the migration and new auth-v2 implementation starts functioning, the user may even create/drop/alter some roles. But then the first coordinator, unaware of failover, commits old data from the migrated tables again.

Reply:
> Do we need to guarantee for auth DML and migration_manager statements to run on topology coordinator only, or are they allowed to be executed on any node?

on any node in group0 as I understand, ok so that's the difference

Reply:
> on any node in group0 as I understand, ok so that's the difference

Precisely, and in contrast, migration needs to be performed in sequence by a single process (topology coordinator), if that process dies and another process continues to a next step, the old process cannot repeat one of the previous steps, as that would introduce inconsistency

Reply:
>  But then the first coordinator, unaware of failover, commits old data from the migrated tables again.

Wouldn't that generate concurrent modification on first coordinator?

Reply:
> if that process dies and another process continues to a next step

And of course another process may continue not because the old process died, but because it got partitioned and the new process decided that the old one is dead. (Failure detection is unreliable in asynchronous networks)

Reply:
> Wouldn't that generate concurrent modification on first coordinator?

That's what a guard protects from. But the problem here is that we're dropping a guard and taking a new one. What happens in between is out of our control.

Reply:
> It doesn't seem that problematic at a glance, you can provide a function that obtains a guard as a parameter to the functions which need to take a guard

but still nothing protects from unaware coder to write some new auth-v2 code which will wrong start_operation and then if it happens to be executed in migration then bang.

I'd either prepare virtual derivative of raft_group0_client (but perhaps that's too much for this single use case) or better add a switch to raft_group0_client which would turn it into "coordinator fiber safe" mode.

Reply:
> better add a switch to raft_group0_client which would turn it into "coordinator fiber safe" mode.

No, you'd be leaking implementation details of higher-level abstraction into a lower-level abstraction. Group 0 is not aware of even the existance of a concept such as "topology coordinator"

Reply:
> but still nothing protects from unaware coder to write some new auth-v2 code which will wrong start_operation and then if it happens to be executed in migration then bang.

The coder should first understand the API they are using and the guarantees it provides (and what it does not) and the context in which they use the API. Not use some API blindly without considering the implications because someone told them to

> I'd either prepare virtual derivative of raft_group0_client (but perhaps that's too much for this single use case) or better add a switch to raft_group0_client which would turn it into "coordinator fiber safe" mode.

How would the coder know which type of client / which switch to use? Same problem, if they are unaware, they will make a mistake. Hence we have review

Reply:
> No, you'd be leaking implementation details of higher-level abstraction into a lower-level abstraction

But why auth code should be aware of topology coordinator? It's the same but from the other side.

Reply:
> But why auth code should be aware of topology coordinator? It's the same but from the other side.

But here we're talking about migration code that is literally part of topology coordinator implementation. The `topology_coordinator::build_coordinator_state` function. Of course it has to be aware

Reply:
> That's what a guard protects from. But the problem here is that we're dropping a guard and taking a new one

You see now you're solving the problem you've created! Or rather not solved in right layer. I wrote here already that it'd be beneficial to support multiple commands under the same guard - looks like it's the same problem I am facing in auth. Why topology migration can't use a single guard? Because it commits multiple commands... Hence you patch it with _term check. 

> you'd be leaking implementation details of higher-level abstraction into a lower-level abstraction

That's not necessary. If you add extra guarantee, which can be turned on in the client, about assuring that all operations span only a single term - that's a knowledge level group0 client already operates on.

>But here we're talking about migration code that is literally part of topology coordinator implementation

Yes and no. Code should be  easily composable and migration is composed out of common methods auth-v2 DML uses. And as I understand you propose introducing specific code or extra abstraction to those common methods.


> if that process dies and another process continues to a next step, the old process cannot repeat one of the previous steps, as that would introduce inconsistency

if we talk in the context of build_coordinator_state function as I understand it does repeat steps, no? When coordinator changes the whole function is executed elsewhere.

> That's what a guard protects from. But the problem here is that we're dropping a guard and taking a new one. What happens in between is out of our control.

I'd like to better understand this _term check.  _term is saved at the beginning so in the following sequence:
1. topology migration starts
2. auth migration starts
3. auth migration finishes
4. migration calls start_operation with term check
5. migration flow continues

if another coordinator is somehow created between steps 2. and 3. (code introduced by me) then at step 4. the migration will be aborted the same way as it is coded today on master. So I fail to understand why if it's safe on master it is not safe after this PR? What am I missing?

> Not so fast, probably not, unless we implement truncate as a group 0 command. But maybe. I don't know.

Right, I guess I'd need to announce it via raft somehow.

Reply:
ok, I understand it now, the guard protection starts from the point it's obtained but any modifications done before would not be caught if not the _term check - which I suppose doesn't guard against all potential modifications but some scenarios around change/duplication of coordinator fiber to a different node (assuming that it's possible to apply multiple commands before the term changes but would need to dig deeper to confirm this). 

> The problem that you're facing now is to perform a sequence of read-modify-writes... atomically. And that's a strictly harder problem, essentially. And the abstraction that we have for solving it is guess what --- the topology coordinator.

Looks like yes, the point is that topology coordinator is lacking things and it likely should not be the place to solve `sequence of read-modify-writes` problem. As it's name suggests it focuses on topology, auth or other features may have nothing to do with topology.  And it still doesn't allow my normal auth code to combine multiple commands under single guard.

Let at least agree that the situation around `start_operation` is not great, I will code it the way it is because I anyway can't do multiple refactors in one PR but:
having `raft_group0_client` with:
```
- start_operation()
- prepare_command()
- add_entry()
```
and having `topology_coordinator` with
```
- start_operation()
```
which I need to carry as a parameter (function pointer or sth) to my feature's logic so that it can call
2 functions from `raft_group0_client` and 1 through that pointer otherwise my code blows up under very  rare circumstances is not great.

Reply:
> And it still doesn't allow my normal auth code to combine multiple commands under single guard

The guard is a kind of a lock.  Using a guard frees the lock.  Just replace `guard`  with `lock` each time you write it and you will see how little sense it makes to ask what you ask for.  Basically what you are asking for is to be able to take and release a lock and still some how guaranty that operations you make are atomic. Unless only a single thread changes the state it is impossible to guaranty. And the topology coordinator is such a thread. And the solution works only if having only one thread that changes the state is acceptable. Otherwise you need to write everything under the same lock. Doing upgrade is as single thread is OK. Executing on auth queries on a single thread is not OK.

Reply:
> Executing on auth queries on a single thread is not OK

It is ok. We talk about change of password or change of role permissions. We don't need to scale it to 1k op/s. 

> The guard is a kind of a lock. Using a guard frees the lock

This makes sense if you look it from one angle. But then imagine that guard is a mutex kind of lock. And imagine each cpu instruction frees your mutex, still makes sense? 

Guard itself is ok abstraction, for some purposes, but for some you need something more (e.g. different type of guard or sth).

I understand, after this discussion even better that topology coordinator is some kind of solution, I understand that it's in the making, something relatively new. But why it is so unusable for other clients and you say it's perfect I don't understand.

Reply:
> > Executing on auth queries on a single thread is not OK
> 
> It is ok. We talk about change of password or change of role permissions. We don't need to scale it to 1k op/s.

Then make one thread (in the entire cluster) (like topology coordinator but to execute authentication statement) and make sure that all auth requests are forwarded to it. Then you will be able to commit your changes in multiple commands and still be atomic. Good luck.

> 
> > The guard is a kind of a lock. Using a guard frees the lock
> 
> This makes sense if you look it from one angle. But then imagine that guard is a mutex kind of lock. And imagine each cpu instruction frees your mutex, still makes sense?

Is guard released on each instruction? No, you can access the state how much you want while holding it. You do not like lock analogy then think about it as "start transaction" (take the guard)  "end transaction" (write back while releasing the guard).

And about your mutex analogy, the funny thing is this is kind of how CPU actually works" https://en.wikipedia.org/wiki/Load-link/store-conditional. Notice the striking similarity to our guard.

> 
> Guard itself is ok abstraction, for some purposes, but for some you need something more (e.g. different type of guard or sth).
> 
> I understand, after this discussion even better that topology coordinator is some kind of solution, I understand that it's in the making, something relatively new. But why it is so unusable for other clients and you say it's perfect I don't understand.

Nothing is perfect, but unless you have a solution there is not point complaining.

Reply:
>And about your mutex analogy, the funny thing is this is kind of how CPU actually works

That's good analogy, and people still implemented mutex  :)

Solution would be to write topology coordinator in a way that it can be used as an api (like group0 client) rather than putting code directly inside it. Why it knows about auth, cdc, tablets? All those are separate features, they are not topology related themselves. Anyway maybe  exact relation is arguable or not. Main point is that it would be easier anyway to write features isolated by better api boundaries.

Reply:
> > And about your mutex analogy, the funny thing is this is kind of how CPU actually works
> 
> That's good analogy, and people still implemented mutex :)

Well, you are welcome to implement distributed locking using our raft abstraction. I am sure there is a lot of material online.

> 
> Solution would be to write topology coordinator in a way that it can be used as an api (like group0 client) rather than putting code directly inside it.

You do not have to. Topology coordinator authors just nice enough to let you use it so you will not have to write something much more complicated. It is also possible to provide an API to register things to run during upgrade. but this is literally one time thing. The code will be used once to upgrade from non topology coordinator aware clusters to a topology aware clusters (which is a one time thing in the cluster's life). @kostja pushing for auth to be done now exactly because it allows us to piggiback on the upgrade path. Otherwise we will have to write something much more complicated and yes, outside of the topology coordinator. You are taking blessing as a curse :)

> Why it knows about auth, cdc, tablets? 

CDC are tightly coupled with the topology. Auth it does not care about whatsoever, but it was decided to use the same mechanism that runs topology coordinator thread to run upgrade thread as well. Since you want to provide upgrade path to the auth you need to add code to the upgrade path. You add nothing to the topology coordinator.